sstables/compress.cc

```diff
@@ -459,6 +459,203 @@ class compressed_file_data_source_impl : public data_source_impl {
+        }
+
+        if (!_input_stream) {
+            _input_stream = co_await _stream_creator();
+        }
+
+        auto info = get_chunk_info(_current_chunk_index, _pos);
+
+        if (_pos != _beg_pos && info.offset != 0) {
+            throw std::runtime_error(format("compressed reader not aligned to chunk boundary: pos={} offset={}", _pos, info.offset));
+        }
+
+        if (!info.chunk_len) {
+            throw sstables::malformed_sstable_exception(format("compressed raw reader chunk_len must be greater than zero, pos={}", _pos));
+        }
+
+        auto res_units = co_await _permit.request_memory(info.chunk_len);
+        auto buf = co_await _input_stream->read_exactly(info.chunk_len);
+        if (buf.size() != info.chunk_len) {
+            throw sstables::malformed_sstable_exception(format("compressed raw reader hit premature end-of-file at file offset {}, expected chunk_len={}, actual={}", _pos, info.chunk_len, buf.size()));
+        }
+
+        auto compressed_len = info.chunk_len - 4;
+        auto expected_checksum = read_be<uint32_t>(buf.get() + compressed_len);
+        auto actual_checksum = crc32_utils::checksum(buf.get(), compressed_len);
+        if (expected_checksum != actual_checksum) {
+            throw sstables::malformed_sstable_exception(format("compressed chunk of size {} at file offset {} failed checksum, expected={}, actual={}", info.chunk_len, _pos, expected_checksum, actual_checksum));
+        }
+
+        if constexpr (check_digest) {
+            if (_digests.can_calculate_digest) {
+                _digests.actual_digest = checksum_combine_or_feed<crc32_utils>(_digests.actual_digest, actual_checksum, buf.get(), compressed_len);
+                uint32_t be_actual_checksum = cpu_to_be(actual_checksum);
+                _digests.actual_digest = crc32_utils::checksum(_digests.actual_digest,
+                        reinterpret_cast<const char*>(&be_actual_checksum), sizeof(be_actual_checksum));
+            }
+        }
+
+        // Removes [0, info.offset). On the first call _pos may be not aligned to chunk boundary.
+        buf.trim_front(info.offset);
+        _current_chunk_index++;
+        _pos += buf.size();
+
+        if constexpr (check_digest) {
+            if (_digests.can_calculate_digest
+                    && _current_chunk_index == _compression_metadata->offsets.size()
+                    && _digests.expected_digest != _digests.actual_digest) {
+                throw sstables::malformed_sstable_exception(seastar::format("Digest mismatch: expected={}, actual={}", _digests.expected_digest, _digests.actual_digest));
+            }
+        }
+
+        co_return make_tracked_temporary_buffer(std::move(buf), std::move(res_units));
+    }
+
+    virtual future<> close() override {
+        if (!_input_stream) {
+            return make_ready_future<>();
+        }
+        return _input_stream->close();
+    }
+
+    virtual future<temporary_buffer<char>> skip(uint64_t n) override {
```

Comment:
Does this even have to be implemented? This source impl is currently only used to read the entire content of the file. Hence this code is currently never run and not tested. How much can this impl be simplified if we remove all code which is needed to support random access reads?